msal/src/Microsoft.Identity.Client/CacheV2/Impl/FileSystemCredentialPathManager.cs

```diff
@@ -0,0 +1,166 @@
+﻿// ------------------------------------------------------------------------------
+// 
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+// 
+// This code is licensed under the MIT License.
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using Microsoft.Identity.Client.CacheV2.Impl.Utils;
+using Microsoft.Identity.Client.CacheV2.Schema;
+using Microsoft.Identity.Core;
+
+namespace Microsoft.Identity.Client.CacheV2.Impl
+{
+    internal class FileSystemCredentialPathManager : ICredentialPathManager
+    {
+        private const string UserPrefix = "u_";
+        private const string EnvironmentPrefix = "e_";
+        private const string RealmPrefix = "r_";
+        private const string ClientIdPrefix = "c_";
+        private const string FamilyIdPrefix = "f_";
+        private const string UserDataFolder = "UD";
+        private const string AccountsFolder = "Accounts";
+        private const string IdTokensFolder = "ID";
+        private const string AccessTokensFolder = "AT";
+        private const string RefreshTokensFolder = "RT";
+        private const string FamilyRefreshTokensFolder = "FRT";
+        private const string AppMetadataFolder = "AppMetadata";
+        private const string FileExtension = ".bin";
+
+        public string GetCredentialPath(Credential credential)
+        {
+            return GetCredentialPath(
+                credential.HomeAccountId,
+                credential.Environment,
+                credential.Realm,
+                credential.ClientId,
+                credential.FamilyId,
+                credential.CredentialType);
+        }
+
+        public string ToSafeFilename(string data)
```

Comment:
I wonder why this is required and what is the likelyhood / impact name of collisions that this brings.

Reply:
Valentin did a bunch of analysis with the base32 hex namespace and calculated out the collision rates.  I'll see if i can find anything about it and add it in as comments.  It's a crazy rare occurrence in the normal case of a few accounts and many products.